fdbserver/ptxn/test/RealTLog.actor.cpp

```diff
@@ -0,0 +1,274 @@
+	pTLogContext->tagProcessID = processID;
+	pTLogContext->tLogID = deterministicRandom()->randomUniqueID();
+	pTLogContext->persistentQueue = openDiskQueue(pTLogDriverContext->diskQueueBasename,
+	                                              tLogOptions.diskQueueExtension,
+	                                              pTLogContext->tLogID,
+	                                              DiskQueueVersion::V1);
+	pTLogContext->persistentData = keyValueStoreMemory(tLogOptions.kvStoreFilename,
+	                                                   pTLogContext->tLogID,
+	                                                   tLogOptions.kvMemoryLimit,
+	                                                   "fdr",
+	                                                   KeyValueStoreType::MEMORY_RADIXTREE);
+	pTLogDriverContext->pTLogContext = pTLogContext;
+
+	// prepare tLog construction.
+	Standalone<StringRef> machineID = LiteralStringRef("machine");
+	LocalityData localities(
+	    Optional<Standalone<StringRef>>(), pTLogDriverContext->zoneID, machineID, pTLogDriverContext->dcID);
+	localities.set(LiteralStringRef("datacenter"), pTLogDriverContext->dcID);
+	Reference<AsyncVar<ServerDBInfo>> dbInfoRef = makeReference<AsyncVar<ServerDBInfo>>(pTLogDriverContext->dbInfo);
+	Reference<AsyncVar<bool>> isDegraded = FlowTransport::transport().getDegraded();
+	Reference<AsyncVar<UID>> activeSharedTLog(new AsyncVar<UID>(pTLogContext->tLogID));
+	state PromiseStream<InitializeTLogRequest> promiseStream = PromiseStream<InitializeTLogRequest>();
+	Promise<Void> oldLog;
+	Promise<Void> recovery;
+
+	// construct tLog.
+	state Future<Void> tl = tLog(pTLogContext->persistentData,
+	                             pTLogContext->persistentQueue,
+	                             dbInfoRef,
+	                             localities,
+	                             promiseStream,
+	                             pTLogContext->tLogID,
+	                             pTLogDriverContext->workerID,
+	                             false, /* restoreFromDisk */
+	                             oldLog,
+	                             recovery,
+	                             pTLogDriverContext->diskQueueBasename,
+	                             isDegraded,
+	                             activeSharedTLog);
+
+	// start tlog.
+	state InitializeTLogRequest initTlogReq = InitializeTLogRequest();
+	initTlogReq.isPrimary = true;
+	TLogInterface interface = wait(promiseStream.getReply(initTlogReq));
+	pTLogContext->realTLogInterface = interface;
+
+	// inform other actors tLog is ready.
+	pTLogContext->TLogCreated.send(true);
+
+	// wait for either test completion or abnormal failure.
+	choose {
+		when(wait(tl)) {}
+		when(bool testCompleted = wait(pTLogContext->realTLogTestCompleted.getFuture())) {}
+	}
+
+	return Void();
+}
+
+// sent commits through TLog interface.
+ACTOR Future<Void> getTLogCommitActor(std::shared_ptr<TestDriverContext> pTestDriverContext,
```

Comment:
To me, this actor name can be something like `sendCommitMessages`, and it can be a member function of the `TLogDriverContext`. The trick is to declare two functions, e.g.,

```
class TLogDriverContext {
...
    // Note this is a member function
    Future<Void> sendCommitMessage(std::shared_ptr<TestDriverContext> pTestDriverContext) {
        return sendCommitMessage_impl(pTestDriverContext, this);
    }

    // this is a static ACTOR function
    ACTOR static Future<Void> sendCommitMessage_impl(std::shared_ptr<TestDriverContext> pTestDriverContext, TLogDriverContext* pTLogDriverContext) {
        // the actual implementation
    }
};